extension/observer/cfgardenobserver/extension.go

```diff
@@ -4,33 +4,337 @@
+	labels := make(map[string]string)
+	tags, err := parseTags(info)
+	if err != nil {
+		g.logger.Warn("not able to parse container tags into labels", zap.Error(err))
+		return nil
+	}
+	for k, v := range tags {
+		labels[k] = v
+	}
+
+	if app != nil {
+		for k, v := range app.Metadata.Labels {
+			labels[k] = *v
+		}
+	}
+
+	return labels
+}
+
+// The info.Properties contains a key called "log_config", which
+// has contents that look like the following JSON endoded string:
+//
+//	{
+//	  "guid": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	  "index": 0,
+//	  "source_name": "CELL",
+//	  "tags": {
+//	    "app_id": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	    "app_name": "example-app",
+//	    "instance_id": "0",
+//	    "organization_id": "11111111-2222-3333-4444-555555555555",
+//	    "organization_name": "example-org",
+//	    "process_id": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	    "process_instance_id": "abcdef12-3456-7890-abcd-ef1234567890",
+//	    "process_type": "web",
+//	    "source_id": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	    "space_id": "99999999-8888-7777-6666-555555555555",
+//	    "space_name": "example-space"
+//	  }
+//	}
+//
+// We parse only the tags into a map, to be used as labels
+func parseTags(info garden.ContainerInfo) (map[string]string, error) {
+	logConfig, ok := info.Properties["log_config"]
+	if !ok {
+		return nil, errors.New("container properties do not have log_config field")
+	}
+
+	var data map[string]any
+	err := json.Unmarshal([]byte(logConfig), &data)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling logConfig: %w", err)
+	}
+
+	tags, ok := data["tags"].(map[string]any)
+	if !ok {
+		return nil, fmt.Errorf("expected tags field to be a map. got=%T", data["tags"])
+	}
+
+	result := make(map[string]string)
+	for key, value := range tags {
+		if strValue, ok := value.(string); ok {
+			result[key] = strValue
+		}
+	}
+
+	return result, nil
+}
+
+func newCfClient(cfConfig CfConfig) (*client.Client, error) {
+	var cfg *config.Config
+	var err error
+
+	switch cfConfig.AuthType {
+	case AuthTypeUserPass:
+		cfg, err = config.New(cfConfig.Endpoint, config.UserPassword(cfConfig.Username, cfConfig.Password))
+	case AuthTypeClientCredentials:
+		cfg, err = config.New(cfConfig.Endpoint, config.ClientCredentials(cfConfig.ClientID, cfConfig.ClientSecret))
+	case AuthTypeToken:
+		cfg, err = config.New(cfConfig.Endpoint, config.Token(cfConfig.AccessToken, cfConfig.RefreshToken))
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error creating connection to CloudFoundry API: %w", err)
+	}
+
+	c, err := client.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (g *cfGardenObserver) updateContainerCache(infos map[string]garden.ContainerInfo) {
+	g.containerMu.Lock()
+	defer g.containerMu.Unlock()
+	g.containers = infos
+}
+
+func infoToAppID(info garden.ContainerInfo) (string, error) {
```

Comment:
I don't really think having this in a separate function provides much value. Callers still have to check the returned error, so it just ends up being a function call vs. accessing the map directly.